internal/plugin/util/converters.go

```diff
@@ -9,41 +9,58 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/types"
 )
 
-// Unmarshal unmarshals a value from in to out.
+// remarshal unmarshals a value from in to out.
 // A rebranded copy from SDK package to avoid circular dependency.
-func Unmarshal(in, out any) error {
+func remarshal(in, out any) error {
 	b, err := json.Marshal(in)
 	if err != nil {
 		return err
 	}
 	return json.Unmarshal(b, out)
 }
 
-// CastSlice casts a slice of any to a slice of T.
-func CastSlice[T any](v any) ([]T, error) {
-	items, ok := v.([]any)
-	if !ok {
-		return nil, fmt.Errorf("expected []any, got %T", v)
+// Unmarshal unmarshals a value from in to out, applies typed modifiers before unmarshalling to out.
+func Unmarshal[I any, O any](in *I, out *O, modifiers ...MapModifier[I]) error {
+	if len(modifiers) == 0 {
+		return remarshal(in, out)
 	}
 
-	result := make([]T, 0, len(items))
-	for _, item := range items {
-		v, ok := item.(T)
-		if !ok {
-			return nil, fmt.Errorf("expected %T element, got %T", v, item)
+	// To applied modifiers, we need to unmarshal into a map[string]any
+	m := make(map[string]any)
+	err := remarshal(in, &m)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal map: %w", err)
+	}
+
+	for _, modify := range modifiers {
+		err := modify(m, in)
+		if err != nil {
+			return fmt.Errorf("failed to apply modifier: %w", err)
 		}
-		result = append(result, v)
 	}
-	return result, nil
+
+	return remarshal(&m, out)
 }
 
-// Expand sets values to Request from a nested attribute/block
+// MapModifier modifies Request and Response objects
+// reqRsp — a map that would be used to unmarshal a Request/Response object
+// in *I — input type, for instance, a dataModel or foo.CreateIn dto.
+// Hint: to ignore "in" type and modify only the map, use `any`, for instance:
+//
+//	func foo[T any](reqRsp map[string]any, _ *T) error {
+//		reqRsp["foo"] = "bar"
+//	}
+//
+// In this case, the function can be used with any type of Request/Response object.
+type MapModifier[I any] func(reqRsp map[string]any, in *I) error
```

Comment:
This is the coolest part: the map can be modified and the `in` would be `dtoModel` or gen-client response (depends on the use-case).

This allows us to convert emails to a list of strings or to a list of maps.